File: app/src/main/java/com/vincent/githubusersearch/viewmodel/UserViewModel.java
package com.vincent.githubusersearch.viewmodel;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import androidx.paging.PagedList;

import com.vincent.githubusersearch.callbacks.PagingStatusCallback;
import com.vincent.githubusersearch.model.User;
import com.vincent.githubusersearch.paging.GitHubUserDataRepo;

public class UserViewModel extends ViewModel {

    private PagingStatusCallback statusCallback;

    public final MutableLiveData<Boolean> liveLoadingStatus = new MutableLiveData<>();
    public final MutableLiveData<String> liveErrorMessage = new MutableLiveData<>();

    public void setStatusCallback(PagingStatusCallback statusCallback) {
        this.statusCallback = statusCallback;
    }

    public LiveData<PagedList<User>> getUserList(String keyword) {
        return new GitHubUserDataRepo(keyword, statusCallback).getUserList();
    }

    public void getData() {
        if (!isLoading()) {
            statusCallback.getData();
        }
    }

    private boolean isLoading() {
        return liveLoadingStatus.getValue() != null && liveLoadingStatus.getValue();
    }
}

File: app/src/main/java/com/vincent/githubusersearch/viewmodel/UserViewModel.kt
package com.vincent.githubusersearch.viewmodel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.paging.PagedList
import com.vincent.githubusersearch.callbacks.PagingStatusCallback
import com.vincent.githubusersearch.model.User
import com.vincent.githubusersearch.paging.UserDataRepo

class UserViewModel : ViewModel() {

    private var statusCallback: PagingStatusCallback? = null

    val liveLoadingStatus = MutableLiveData<Boolean>()
    val liveErrorMessage = MutableLiveData<String>()

    fun setStatusCallback(statusCallback: PagingStatusCallback) {
        this.statusCallback = statusCallback
    }

    fun getUserList(keyword: String): LiveData<PagedList<User>> {
        return UserDataRepo(keyword, statusCallback).userList
    }

    fun getData() {
        if (!isLoading()) {
            statusCallback?.requireData()
        }
    }

    private fun isLoading(): Boolean {
        return liveLoadingStatus.value ?: false
    }
}